### PR TITLE
add `.abc` and `.typing` modules

### LINKED ISSUE #1634: 0.11 API
# new top-level

- [x] #1644
- [x] #1641

# .experimental
 
- [x] #1642
- [x] `experimental.read_elem_as_dask` → `experimental.read_elem_lazy`? Not yet: https://github.com/scverse/anndata/issues/1634#issuecomment-2325894399
- [x] #1643

# new modules

- [x] #1645

### PR DESCRIPTION
**Pull request by flying-sheep**
Fixes #1634

TODO:

- [x] Add actual API we want to commit to to `abc.*`

### CODE REVIEW

**Comment by ilan-gold** [comment]
> Add actual API we want to commit to to abc.*

I made an attempt at what I thought was reasonable via `BaseCompressedSparseDataset`'s private/`_` methods but I would also be ok with a first pass at this rolling that back further to not include `append`, `__setitem__`, or `group` i.e., people should be able to "get" (either `__getitem__` or `to_memory()`),  ask for the backing type (`zarr` or `hdf5`), know the `dtype`, and get the shape.

**Comment by ilan-gold** [comment]
After all, easier to add than remove...

**Comment by flying-sheep** [comment]
We could export the `Index` type for `__getitem__` or some simplification.

**Comment by ilan-gold** [comment]
> We could export the Index type for __getitem__ or some simplification.

Absolutely.  Maybe a `types` module then here as well?

**Comment by ilan-gold** [comment]
Either that, or punt on `types` until 0.12.  I don't have a preference there.

File: src/anndata/abc.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    shape: tuple[int, int]
+    """Shape of the matrix."""
+
+    dtype: np.dtype
+    """The :class:`numpy.dtype` of the `data` attribute of the sparse matrix."""
+
+    backend: Literal["zarr", "hdf5"]
+    """Which file type is used on-disk."""
+
+    # TODO: index type
```
Comment (ilan-gold) [inline src/anndata/abc.py:33]: Export that in `types`? I noticed it in a few spots in the github search for `_code`

File: src/anndata/typing.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    AwkArray,
+    H5Array,
+    ZarrArray,
+    ZappyArray,
+    abc.CSRDataset,
+    abc.CSCDataset,
+    DaskArray,
+    CupyArray,
+    CupySparseMatrix,
+]
```
Comment (ilan-gold) [inline src/anndata/typing.py:45]: What's supposed to be the difference between this file and `_types` if this isn't public?

**Review by ilan-gold — CHANGES_REQUESTED** [review]
Please look at #1647 or fix CI some other way!

File: src/anndata/abc.py (hunk 1 above)
Comment (flying-sheep) [reply src/anndata/abc.py:33]: Do you mean `_types` and `_core`?

**Comment by flying-sheep** [comment]
> Please look at https://github.com/scverse/anndata/pull/1647 or fix CI some other way!

whoop, seems like we did the same thing!

**Comment by ilan-gold** [comment]
> whoop, seems like we did the same thing!

Haha, but yours is much cleaner!

File: src/anndata/typing.py (hunk 2 above)
Comment (flying-sheep) [reply src/anndata/typing.py:45]: Everything in here is public, just not all names.

There are a few internal APIs that use subsets of the public types, and these are included in here too.

I’d have put these into `_types` if Python was able to deal with this kind of circularity.

So I guess the logic is “all public types and the types they depend on?

**Comment by flying-sheep** [comment]
I wouldn’t say so, I just simplified the error message again (after making it more complex in another PR :see_no_evil:)

File: src/anndata/abc.py (hunk 1 above)
Comment (ilan-gold) [reply src/anndata/abc.py:33]: Sorry! `typing`!

Comment (ilan-gold) [reply src/anndata/abc.py:33]: So we should export `Index` in `typing`

File: src/anndata/typing.py (hunk 2 above)
Comment (ilan-gold) [reply src/anndata/typing.py:45]: Sounds good!

**Review by ilan-gold — APPROVED** [review]

**Pull request merged**